src/compiler/parser.ts

```diff
@@ -1001,10 +1001,10 @@ module ts {
             switch (parsingContext) {
                 case ParsingContext.SourceElements:
                 case ParsingContext.ModuleElements:
-                    return isSourceElement(inErrorRecovery);
+                    return !(token === SyntaxKind.SemicolonToken && inErrorRecovery) && isModuleElement();
                 case ParsingContext.BlockStatements:
                 case ParsingContext.SwitchClauseStatements:
-                    return isStartOfStatement(inErrorRecovery);
+                    return !(token === SyntaxKind.SemicolonToken && inErrorRecovery) && isStatement();

```

Comment:
No, it's better this way. Plus there's no reason why `isStatement` should be bothered with an `inErrorRecovery` flag that it doesn't otherwise need.